File: src/launcher/SharpBuy_Launcher_Mac/src-tauri/src/accounts.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

pub const MAX_ACCOUNTS: usize = 500;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct SavedAccount {
    pub steam_id: String,
    pub account_name: String,
    pub persona_name: String,
    pub avatar_url: String,
    pub token: String,
    pub added_at: i64,
    pub warranty_expires_at: i64,
    pub exp_seconds: i64,
    #[serde(default)]
    pub last_checked_at: i64,
    pub is_alive: bool,
    pub status_message: String,
    pub vac_banned: String,
}

pub fn app_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("Library/Application Support/SharpBuy_Launcher")
}

pub fn accounts_path() -> PathBuf {
    app_dir().join("accounts.json")
}

fn ensure_dir() -> std::io::Result<()> {
    fs::create_dir_all(app_dir())
}

pub fn load_accounts() -> Vec<SavedAccount> {
    ensure_dir().ok();
    let path = accounts_path();
    if !path.exists() {
        return vec![];
    }
    fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default()
}

pub fn save_accounts(accounts: &[SavedAccount]) -> Result<(), String> {
    ensure_dir().map_err(|e| e.to_string())?;
    let json = serde_json::to_string_pretty(accounts).map_err(|e| e.to_string())?;
    fs::write(accounts_path(), json).map_err(|e| e.to_string())
}

pub fn get_saved_accounts_json() -> String {
    serde_json::to_string(&load_accounts()).unwrap_or_else(|_| "[]".to_string())
}

pub fn upsert_account(entry: SavedAccount) -> Result<(), String> {
    let mut accounts = load_accounts();
    if let Some(idx) = accounts.iter().position(|a| a.steam_id == entry.steam_id) {
        accounts[idx] = entry;
    } else {
        if accounts.len() >= MAX_ACCOUNTS {
            accounts.remove(0);
        }
        accounts.push(entry);
    }
    save_accounts(&accounts)
}

pub fn delete_account(steam_id: &str) -> Result<(), String> {
    let accounts: Vec<_> = load_accounts()
        .into_iter()
        .filter(|a| a.steam_id != steam_id)
        .collect();
    save_accounts(&accounts)
}

pub fn clear_all() -> Result<(), String> {
    save_accounts(&[])
}

pub fn account_from_json_value(v: &Value) -> Option<SavedAccount> {
    serde_json::from_value(v.clone()).ok()
}

File: src/launcher/SharpBuy_Launcher_Electron/src/accounts.js
const fs = require('fs');
const path = require('path');
const os = require('os');

const MAX_ACCOUNTS = 500;

function appDir() {
  return path.join(os.homedir(), 'Library/Application Support/SharpBuy_Launcher');
}

function accountsPath() {
  return path.join(appDir(), 'accounts.json');
}

function gamesCachePath() {
  return path.join(appDir(), 'games_cache.json');
}

function ensureDir() {
  fs.mkdirSync(appDir(), { recursive: true });
}

function loadAccounts() {
  ensureDir();
  const file = accountsPath();
  if (!fs.existsSync(file)) return [];
  try {
    const data = JSON.parse(fs.readFileSync(file, 'utf8'));
    return Array.isArray(data) ? data : [];
  } catch (_) {
    return [];
  }
}

function saveAccounts(accounts) {
  ensureDir();
  fs.writeFileSync(accountsPath(), JSON.stringify(accounts, null, 2), 'utf8');
}

function getSavedAccountsJson() {
  return JSON.stringify(loadAccounts());
}

function upsertAccount(entry) {
  const accounts = loadAccounts();
  const idx = accounts.findIndex((a) => a.SteamId === entry.SteamId);
  if (idx !== -1) {
    accounts[idx] = entry;
  } else {
    if (accounts.length >= MAX_ACCOUNTS) {
      accounts.shift();
    }
    accounts.push(entry);
  }
  saveAccounts(accounts);
}

function deleteAccount(steamId) {
  const accounts = loadAccounts().filter((a) => a.SteamId !== steamId);
  saveAccounts(accounts);
}

function clearAll() {
  saveAccounts([]);
}

function getCachedGamesJson() {
  ensureDir();
  const file = gamesCachePath();
  if (!fs.existsSync(file)) return '{}';
  try {
    const raw = fs.readFileSync(file, 'utf8');
    JSON.parse(raw);
    return raw;
  } catch (_) {
    return '{}';
  }
}

function saveCachedGames(json) {
  ensureDir();
  try {
    JSON.parse(json || '{}');
  } catch (_) {
    return;
  }
  fs.writeFileSync(gamesCachePath(), json, 'utf8');
}

function makeAccount({
  steamId,
  accountName,
  personaName,
  avatarUrl,
  token,
  addedAt,
  warrantyExpiresAt = 0,
  expSeconds,
  lastCheckedAt = 0,
  isAlive = true,
  statusMessage = '',
  vacBanned = '0',
}) {
  return {
    SteamId: steamId,
    AccountName: accountName,
    PersonaName: personaName,
    AvatarUrl: avatarUrl,
    Token: token,
    AddedAt: addedAt,
    WarrantyExpiresAt: warrantyExpiresAt,
    ExpSeconds: expSeconds,
    LastCheckedAt: lastCheckedAt,
    IsAlive: isAlive,
    StatusMessage: statusMessage,
    VacBanned: vacBanned,
  };
}

module.exports = {
  MAX_ACCOUNTS,
  loadAccounts,
  saveAccounts,
  getSavedAccountsJson,
  upsertAccount,
  deleteAccount,
  clearAll,
  getCachedGamesJson,
  saveCachedGames,
  makeAccount,
};
